### PR TITLE
Allow `~const` bounds on trait assoc functions

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk

### CODE REVIEW

File: src/test/ui/rfc-2632-const-trait-impl/trait-where-clause.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#![feature(const_fn_trait_bound)]
+#![feature(const_trait_impl)]
+
+trait Bar {}
+
+trait Foo {
+    fn a();
+    fn b() where Self: ~const Bar;
```
Comment (oli-obk) [inline src/test/ui/rfc-2632-const-trait-impl/trait-where-clause.rs:8]: A few more tests I think we should have:
* `where Self: ~const Foo`
* a const and a nonconst impl for such a trait
* a monomorphic use that actually runs it

File: src/test/ui/rfc-2632-const-trait-impl/trait-where-clause-run.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+}
+
+impl const Foo for Const {}
+
+fn main() {
+    const ANS1: u8 = Const::foo();
+    let ans2 = NonConst::foo();
+
+    assert_eq!(ANS1 + ans2, 42);
```
Comment (oli-obk) [inline src/test/ui/rfc-2632-const-trait-impl/trait-where-clause-run.rs:40]: beautiful

**Comment by nalply** [comment]
I got here by TWIR and I never have seen `~const` in Rust. What's this? Sorry for the noise, `~const` is not googlable.

**Comment by oli-obk** [comment]
it's placeholder syntax for "const if const" trait bounds: https://github.com/rust-lang/rust/issues/67792

the full updated RFC text can be found at https://internals.rust-lang.org/t/pre-rfc-revamped-const-trait-impl-aka-rfc-2632/15192

**Pull request merged**